beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -31,36 +32,191 @@ Introducing the rollback to scaffolder actions provides the mean to come back to
+  ...
+  handler: (
+    ctx: ActionContext<TActionInput, TActionOutput, TSchemaType>,
+  rollback?: (
+    ctx: ActionContext<TActionInput, TActionOutput, TSchemaType>,
+  ) => Promise<void>;
+};
+```
+
+Implementers provide the rollback function when calling `createTemplateAction`.
 
 ```typescript
 const createPublishGitHubAction = createTemplateAction({
   id: 'publish:github',
+  ...,
   async handler() {},
   async rollback() {},
 });
 ```
 
+#### Example
+
+```tsx
+createTemplateAction<{
+  apiUrl: (z) => z.string(),
+  projectKey: (z) => z.string(),
+  // other inputs...
+}, {
+  projectId: (z) => z.string(),
+}>(
+  {
+    ...,
+    handler: async (ctx) => {
+      // Create the project
+      const projectId = await create(ctx.input.apiUrl, ctx.input.projectKey);
+      ctx.output.projectId = projectId;
+    },
+    rollback: async (ctx) => {
+      // Delete the project if it was created
+      if (ctx.output.projectId) {
+        await delete(ctx.input.apiUrl, ctx.output.projectId);
+      }
+    }
+  }
+)
+```
+
+#### Example scenario
+
+The rollback will be performed in the reverse order of the execution. For example, if the following actions are executed, where the rollback is provided for the last 3 actions:
+
+1. create a repository (no rollback provided)
+1. create a pull request (rollback provided)
+1. create a branch (rollback provided)
+1. create third party resource (rollback provided)
+
+The rollback will be performed in the following order:
+
+1. delete the third party resource
+1. delete the branch
+1. delete the pull request
+
+The repository will not be deleted because the rollback is not provided for it.
+
+### WorkflowRunner changes
+
+When a scaffolder task fails, the system will invoke the rollback function for any actions that:
+
+1. Were successfully executed
+1. Provide a rollback implementation
+
+The rollback execution will follow a reverse order (LIFO approach) from the original execution, ensuring dependent resources are cleaned up properly.
+
+The NunjucksWorkflowRunner will be modified to track successfully completed actions with their rollback function.
+
+```typescript
+const completedActionsWithRollback: Array<{
+  action: TemplateAction;
+  ctx: ActionContext<any, any, any>;
+  step: TaskStep;
+}> = [];
+```
+
+```typescript
+const completedActionsWithRollback = [];
+
+async executeStep(
+  task: TaskContext,
+  step: TaskStep,
+  context: TemplateContext,
+  ...
+) {
+  // ...
+  await action.handler(ctx);
+
+  if (action.rollback && !task.isDryRun) {
+    completedActionsWithRollback.push({ action, ctx, step });
+  }
+
+  // ...
+}
+```
+
+In case of a failure, the workflow runner will invoke the rollback function for each of the completed actions.
+
+```ts
+for (const { action, ctx, step } of [
+  ...this.completedActionsWithRollback,
+].reverse()) {
+  if (action.rollback) {
+    // ...
+    taskLogger.info(`Rolling back step ${step.id} (${step.name})`);
+    await action.rollback(ctx);
```

Comment:
Just wondering what happens if the rollback itself fails? Should it be ignored or does it stop the next rollbacks as well?

Reply:
I lean towards having the worker complete as many of the rollbacks as possible, so if one fails, we continue, and give a report at the end. 

In our use case, if we create 5 resources, I'd rather get 4/5 cleaned up rather than 0/5 if one was to fail.

Reply:
Agree. Or maybe this could be configurable and default to that behaviour? But anyways, it probably should be part of this BEP

Reply:
I've updated the content to be more explicit about what happens when a single rollback fails and mentioned the idea of overriding with something like `dangerouslyEnableRollback`.

Reply:
Good job, makes sense!

Reply:
Following up on some other feedback - this would likely be configurable at a `TaskRecoverStrategy` level.